### PR TITLE
[WIP] Added new command "Expand Selection To Brackets."

### PR DESCRIPTION
**Pull request by PxlBuzzard**
This feature is intended to work similarly to Sublime's "Expand Selection to Brackets." This should fulfill #14089.

Features left to complete:

- [x] Handle selections, not just a cursor
- [x] Multi-cursor
- [ ] Tests
- [ ] Code cleanup

### CODE REVIEW

File: src/vs/editor/contrib/smartSelect/common/smartSelect.ts [hunk 1]
```diff
@@ -185,8 +187,8 @@ class ShrinkSelectionAction extends AbstractSmartSelect {
 	constructor() {
 		super(false, {
 			id: 'editor.action.smartSelect.shrink',
-			label: nls.localize('smartSelect.shrink', "Shrink Select"),
-			alias: 'Shrink Select',
+			label: nls.localize('smartSelect.shrink', "Shrink Selection"),
+			alias: 'Shrink Selection',
```
Comment (PxlBuzzard) [inline src/vs/editor/contrib/smartSelect/common/smartSelect.ts]: Side note, I am making this change after talking with @gregvanl about the new feature. This will standard all selection text to use the word "selection."

File: src/vs/editor/contrib/smartSelect/common/smartSelect.ts [hunk 2]
```diff
@@ -196,3 +198,84 @@ class ShrinkSelectionAction extends AbstractSmartSelect {
...
+class SelectBracketAction extends EditorAction {
+	constructor() {
+		super({
+			id: 'editor.action.expandSelectionToBrackets',
+			label: nls.localize('smartSelect.expandSelectionToBrackets', "Expand Selection To Brackets"),
+			alias: 'Expand Selection To Brackets',
+			precondition: null,
+			kbOpts: {
+				kbExpr: EditorContextKeys.textFocus,
+				primary: KeyMod.CtrlCmd | KeyMod.Shift | KeyCode.US_SLASH
```
Comment (PxlBuzzard) [inline src/vs/editor/contrib/smartSelect/common/smartSelect.ts]: Sublime uses Ctrl+Shift+M for this, but we already have that on "Show Problems." Any suggestions?

Comment (HazardDev) [inline src/vs/editor/contrib/smartSelect/common/smartSelect.ts]: You could use Ctrl + K, Shift + M so it's _somewhat_ similar.

Comment (HazardDev) [inline src/vs/editor/contrib/smartSelect/common/smartSelect.ts]: Although on further thought this would be a bit of a hassle to press a few times... Maybe something as basic as `Alt + M` would work.

Comment (chrmarti) [inline src/vs/editor/contrib/smartSelect/common/smartSelect.ts]: `Alt` keybindings don't work well on Linux.

File: src/vs/editor/contrib/smartSelect/common/smartSelect.ts [hunk 3]
```diff
@@ -196,3 +198,84 @@ class ShrinkSelectionAction extends AbstractSmartSelect {
...
+
+		const selection = editor.getSelection();
+		if (!selection.isEmpty()) {
+			return;
+		}
+
+		const originalPosition = selection.getStartPosition();
+
+		// find previous open bracket with no match
+		let prevBracket = null;
```
Comment (chrmarti) [inline src/vs/editor/contrib/smartSelect/common/smartSelect.ts]: Add type.

File: src/vs/editor/contrib/smartSelect/common/smartSelect.ts [hunk 4]
```diff
@@ -196,3 +198,84 @@ class ShrinkSelectionAction extends AbstractSmartSelect {
...
+		let currentPosition = new Position(originalPosition.lineNumber, originalPosition.column);
```
Comment (chrmarti) [inline src/vs/editor/contrib/smartSelect/common/smartSelect.ts]: Do you need to copy the position?

File: src/vs/editor/contrib/smartSelect/common/smartSelect.ts [hunk 5]
```diff
@@ -196,3 +198,84 @@ class ShrinkSelectionAction extends AbstractSmartSelect {
...
+		let setCount = 0;
+		do {
+			prevBracket = model.findPrevBracket(currentPosition);
+			if (prevBracket) {
+				currentPosition = new Position(prevBracket.range.startLineNumber, prevBracket.range.startColumn);
```
Comment (chrmarti) [inline src/vs/editor/contrib/smartSelect/common/smartSelect.ts]: Use prevBracket.range.getStartPosition()

File: src/vs/editor/contrib/smartSelect/common/smartSelect.ts [hunk 6]
```diff
@@ -196,3 +198,84 @@ class ShrinkSelectionAction extends AbstractSmartSelect {
...
+				!prevBracket.isOpen ? setCount++ : setCount--;
+			}
+		} while (prevBracket && !(prevBracket.isOpen && setCount < 0));
+
+		// find next closed bracket with no match
+		let nextBracket = null;
```
Comment (chrmarti) [inline src/vs/editor/contrib/smartSelect/common/smartSelect.ts]: Add type.

File: src/vs/editor/contrib/smartSelect/common/smartSelect.ts [hunk 7]
```diff
@@ -196,3 +198,84 @@ class ShrinkSelectionAction extends AbstractSmartSelect {
...
+		currentPosition = new Position(originalPosition.lineNumber, originalPosition.column);
```
Comment (chrmarti) [inline src/vs/editor/contrib/smartSelect/common/smartSelect.ts]: Copy needed?

File: src/vs/editor/contrib/smartSelect/common/smartSelect.ts [hunk 8]
```diff
@@ -196,3 +198,84 @@ class ShrinkSelectionAction extends AbstractSmartSelect {
...
+		setCount = 0;
+		do {
+			nextBracket = model.findNextBracket(currentPosition);
+			if (nextBracket) {
+				currentPosition = new Position(nextBracket.range.endLineNumber, nextBracket.range.endColumn);
```
Comment (chrmarti) [inline src/vs/editor/contrib/smartSelect/common/smartSelect.ts]: Use range.getEndPosition() (also further down)

File: src/vs/editor/contrib/smartSelect/common/smartSelect.ts (hunk 5 above)
Comment (chrmarti) [inline src/vs/editor/contrib/smartSelect/common/smartSelect.ts]: (also further down)

File: src/vs/editor/contrib/smartSelect/common/smartSelect.ts [hunk 9]
```diff
@@ -196,3 +198,84 @@ class ShrinkSelectionAction extends AbstractSmartSelect {
...
+
+		// check validity of prev and next brackets
+		if (!prevBracket && !nextBracket) {
+			return;
+		}
+
+		// pick the outermost set of brackets
+		const matchPrev = prevBracket ? model.matchBracket(new Position(prevBracket.range.startLineNumber, prevBracket.range.startColumn)) : null;
+		const matchNext = nextBracket ? model.matchBracket(new Position(nextBracket.range.startLineNumber, nextBracket.range.startColumn)) : null;
+		let bracketsToSelect = null;
```
Comment (chrmarti) [inline src/vs/editor/contrib/smartSelect/common/smartSelect.ts]: Add type.

File: src/vs/editor/common/editorCommon.ts [hunk 10]
```diff
@@ -875,12 +875,13 @@ export interface ITokenizedModel extends ITextModel {
...
-	//  * @return The info for the first bracket before `position`, or null if there are no more brackets before `positions`.
-	//  */
-	// findPrevBracket(position:IPosition): IFoundBracket;
+	/**
+	 * Find the first bracket in the model before `position`.
+	 * @param position The position at which to start the search.
+	 * @return The info for the first bracket before `position`, or null if there are no more brackets before `positions`.
+	 * @internal
+	 */
+	findPrevBracket(position:IPosition): IFoundBracket;
```
Comment (chrmarti) [inline src/vs/editor/common/editorCommon.ts]: findNextBracket() below is also needed.

**Review by chrmarti — CHANGES_REQUESTED** [review]

File: src/vs/editor/contrib/smartSelect/common/smartSelect.ts (hunk 2 above)
Comment (HazardDev) [inline src/vs/editor/contrib/smartSelect/common/smartSelect.ts]: `Shift+M` and `Control+M` are both open also.

I, personally, would pick something like `Shift+B` because "B for brackets" is easy to remember.

File: src/vs/editor/contrib/smartSelect/common/smartSelect.ts [hunk 11]
```diff
@@ -196,3 +198,84 @@ class ShrinkSelectionAction extends AbstractSmartSelect {
...
+		if (!matchPrev) {
+			bracketsToSelect = matchNext;
+		} else if (!matchNext) {
+			bracketsToSelect = matchPrev;
+		} else if (!new Position(matchPrev[1].endLineNumber, matchPrev[1].endColumn)
```
Comment (rebornix) [inline src/vs/editor/contrib/smartSelect/common/smartSelect.ts]: Comparing line/column is better than creating object and relying on their comparable methods.

File: src/vs/editor/contrib/smartSelect/common/smartSelect.ts [hunk 12]
```diff
@@ -196,3 +198,84 @@ class ShrinkSelectionAction extends AbstractSmartSelect {
...
+
+		// find previous open bracket with no match
+		let prevBracket = null;
+		let currentPosition = new Position(originalPosition.lineNumber, originalPosition.column);
+		let setCount = 0;
+		do {
+			prevBracket = model.findPrevBracket(currentPosition);
+			if (prevBracket) {
+				currentPosition = new Position(prevBracket.range.startLineNumber, prevBracket.range.startColumn);
+				!prevBracket.isOpen ? setCount++ : setCount--;
```
Comment (rebornix) [inline src/vs/editor/contrib/smartSelect/common/smartSelect.ts]: Doing binary operation against boolean then check whether it's true or false is not intuitive.

File: src/vs/editor/contrib/smartSelect/common/smartSelect.ts [hunk 13]
```diff
@@ -196,3 +198,84 @@ class ShrinkSelectionAction extends AbstractSmartSelect {
...
+			}
+		} while (prevBracket && !(prevBracket.isOpen && setCount < 0));
```
Comment (rebornix) [inline src/vs/editor/contrib/smartSelect/common/smartSelect.ts]: If we use `findPrevBracket` for bracket searching, then it will catch all possible types of brackets `{ [ (` but a single `setCount` can't handle all types.

**Comment by rebornix** [comment]
@PxlBuzzard let's add test cases to cover those edge cases we found this morning first. Then we can have confidence with further code change.

**Comment by rebornix** [comment]
As we didn't plan to have feature in in near future, I'd like to close it for now to help us better maintain issues. We can reopen it when we are going to implement this feature in a proper way. Thanks anyway for your contribution!
